### ISSUE TITLE
[1.3.0] Regression in parsing of nested ADTs

### ISSUE THREAD

**Issue by plokhotnyuk** [sirthias/borer#146]
Here is PR with upgrading of the borer version from `1.2.1` to `1.3.0`: https://github.com/plokhotnyuk/jsoniter-scala/pull/452

One of the nested ADT for GeoJSON cannot be parsed due some unexpected error like here:
```
2020-01-10T15:33:31.6957939Z [info]   io.bullet.borer.Borer$Error$InvalidInputData: Expected type id key for subtype of `com.github.plokhotnyuk.jsoniter_scala.benchmark.GeoJSON.Geometry` but got String (input position 104)
2020-01-10T15:33:31.6958577Z [info]   at io.bullet.borer.InputReader.unexpectedDataItem(Reader.scala:622)
2020-01-10T15:33:31.6958965Z [info]   at io.bullet.borer.InputReader.unexpectedDataItem(Reader.scala:618)
2020-01-10T15:33:31.6959329Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$.fail$3(BorerJsonEncodersDecoders.scala:58)
2020-01-10T15:33:31.6959701Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$.$anonfun$x$9$4(BorerJsonEncodersDecoders.scala:58)
2020-01-10T15:33:31.6960057Z [info]   at io.bullet.borer.InputReader.read(Reader.scala:521)
2020-01-10T15:33:31.6960420Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$FeatureDecoder$1.readObject$14(BorerJsonEncodersDecoders.scala:59)
2020-01-10T15:33:31.6961477Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$FeatureDecoder$1.read(BorerJsonEncodersDecoders.scala:59)
2020-01-10T15:33:31.6962016Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$$anon$28.read(BorerJsonEncodersDecoders.scala:59)
2020-01-10T15:33:31.6962467Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$$anon$28.read(BorerJsonEncodersDecoders.scala:59)
2020-01-10T15:33:31.6962826Z [info]   at io.bullet.borer.InputReader.read(Reader.scala:521)
2020-01-10T15:33:31.6963205Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$.$anonfun$x$9$6(BorerJsonEncodersDecoders.scala:60)
2020-01-10T15:33:31.6963560Z [info]   at io.bullet.borer.InputReader.read(Reader.scala:521)
2020-01-10T15:33:31.6963904Z [info]   at io.bullet.borer.InputReader.rec$1(Reader.scala:525)
2020-01-10T15:33:31.6964250Z [info]   at io.bullet.borer.InputReader.readUntilBreak(Reader.scala:526)
2020-01-10T15:33:31.6964605Z [info]   at io.bullet.borer.Decoder$.$anonfun$forIterable$1(Decoder.scala:255)
2020-01-10T15:33:31.6964956Z [info]   at io.bullet.borer.InputReader.read(Reader.scala:521)
2020-01-10T15:33:31.6965319Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$FeatureCollectionDecoder$1.readObject$15(BorerJsonEncodersDecoders.scala:61)
2020-01-10T15:33:31.6965772Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$FeatureCollectionDecoder$1.read(BorerJsonEncodersDecoders.scala:61)
2020-01-10T15:33:31.6966156Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$$anon$30.read(BorerJsonEncodersDecoders.scala:61)
2020-01-10T15:33:31.6966524Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$$anon$30.read(BorerJsonEncodersDecoders.scala:61)
2020-01-10T15:33:31.6966875Z [info]   at io.bullet.borer.InputReader.read(Reader.scala:521)
2020-01-10T15:33:31.6967239Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.BorerJsonEncodersDecoders$.$anonfun$x$9$8(BorerJsonEncodersDecoders.scala:62)
2020-01-10T15:33:31.6967600Z [info]   at io.bullet.borer.DecodingSetup$Impl.decodeFrom(DecodingSetup.scala:176)
2020-01-10T15:33:31.6967957Z [info]   at io.bullet.borer.DecodingSetup$Impl.value(DecodingSetup.scala:106)
2020-01-10T15:33:31.6968313Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.GeoJSONReading.borerJson(GeoJSONReading.scala:26)
2020-01-10T15:33:31.6968679Z [info]   at com.github.plokhotnyuk.jsoniter_scala.benchmark.GeoJSONReadingSpec.$anonfun$new$2(GeoJSONReadingSpec.scala:9)
```

**Comment by sirthias** [comment]
Ok, thank you.
This is in fact caused by https://github.com/sirthias/borer/issues/145, which I am in the progress of fixing.

**Comment by sirthias** [comment]
Just released borer 1.4.0 which is confirmed to fix this problem in jsoniter-scala.
